stan/math/prim/mat/fun/mrrr.hpp

```diff
@@ -0,0 +1,709 @@
+ * @param[out] min_element_growth Element growth achieved with resulting shift.
+ */
+void find_shift(const Eigen::VectorXd& l, const Eigen::VectorXd& d,
+                const double low, const double high,
+                const double max_ele_growth, const double max_shift,
+                Eigen::VectorXd& l2, Eigen::VectorXd& d2, double& shift,
+                double& min_element_growth) {
+  Eigen::VectorXd l3(l2.size()), d3(d2.size());
+  const std::vector<double> shifts = {
+      low,
+      high - max_shift * 0.1,
+      low + max_shift * 0.1,
+      high - max_shift * 0.25,
+      low + max_shift * 0.25,
+      high - max_shift * 0.5,
+      low + max_shift * 0.5,
+      high - max_shift * 0.75,
+      low + max_shift * 0.75,
+      high - max_shift,
+      low + max_shift,
+  };
+  min_element_growth = std::numeric_limits<double>::infinity();
+  for (double sh : shifts) {
+    const double element_growth = get_shifted_ldl(l, d, sh, l3, d3);
+    if (element_growth < min_element_growth) {
+      l2.swap(l3);
+      d2.swap(d3);
+      shift = sh;
+      min_element_growth = element_growth;
+      if (element_growth <= max_ele_growth) {
+        break;
+      }
+    }
+  }
+}
+
+/**
+ * Finds a good value for shift of the initial LDL factorization T - shift * I =
+ * L * D * L^T.
+ * @param diagonal Diagonal of T.
+ * @param subdiagonal Subdiagonal of T.
+ * @param l0 Subdiagonal of L.
+ * @param d0 Diagonal of D.
+ * @param min_eigval Lower bound on eigenvalues of T.
+ * @param max_eigval High bound on eigenvalues of T
+ * @param max_ele_growth Maximum desired element growth.
+ * @return
+ */
+double find_initial_shift(const Eigen::Ref<const Eigen::VectorXd> diagonal,
+                          const Eigen::Ref<const Eigen::VectorXd> subdiagonal,
+                          Eigen::VectorXd& l0, Eigen::VectorXd& d0,
+                          const double min_eigval, const double max_eigval,
+                          const double max_ele_growth) {
+  double shift = 0;
+  if (min_eigval > 0) {
+    shift = 0.9 * min_eigval;
+  } else if (max_eigval <= 0) {
+    shift = 0.9 * max_eigval;
+  }
+  double element_growth = get_ldl(diagonal, subdiagonal, shift, l0, d0);
+  if (element_growth < max_ele_growth) {
+    return shift;
+  }
+  double plus = (max_eigval - min_eigval) * 1e-15;
+  while (!(element_growth < max_ele_growth)) {  // if condition is flipped it
+                                                // would be wrong for the case
+                                                // where element_growth is nan
+    plus *= -2;
+    element_growth = get_ldl(diagonal, subdiagonal, shift + plus, l0, d0);
+  }
+  return shift + plus;
+}
+
+struct mrrr_task {
+  int start, end;
+  double shift;  // total shift, not just the last one
+  Eigen::VectorXd l, d;
+  int level;
+};
+
+/**
+ * Calculates eigenvalues and eigenvectors of a irreducible tridiagonal matrix T
+ * using MRRR algorithm. Use `tridiagonal_eigensolver` if any subdiagonal
```

Comment:
Can you add a link to the "MRRR algorithm" or at least spell it out once? (Meaning... how do I search and find the algorithm online... I don't know who MRRR is and it's not really getting me anything useful.)